mitodl/bootcamp-ecommerce
# Hubspot CRM API syncing

#### What are the relevant tickets?
Closes #1354 

#### What's this PR do?
Refactors all the old Hubspot code, based on the Ecommerce Bridge API, to work with the CRM API instead.

#### How should this be manually tested?
- If you don't already have existing users, bootcamps, and applications, create some on the master branch. (You can run `manage.py seed_data` to create some test bootcamps).  Then switch to this branch.
- Run `docker-compose build`
- Set `MITOL_HUBSPOT_API_ID_PREFIX ` to something like "bootcamps-<your_initials>-2022" in your .env file
- Set `HUBSPOT_PIPELINE_ID=75e28846-ad0d-4be2-a027-5e1da6590b98`
- Ask me for the value of `MITOL_HUBSPOT_API_PRIVATE_TOKEN` 
- Ask me for access to the hubspot sandbox account
- Start up your docker containers
- Run `manage.py configure_hubspot_properties`, it should complete without errors.  This command upserts custom properties for objects in hubspot.
- Assuming you have some existing users, bootcamps, and applications, try running `manage.py sync_db_to_hubspot create`, it should complete without errors. 
- Check that you have new `HubspotObject` objects, with the appropriate content_type/object_id values for all your users, products, and orders.
- Delete all your `HubspotObject` objects, then run `manage.py sync_hubspot_ids_to_db`. It should find matches for everything and all the `HubspotObject` objects should be back.
- Register as a new user and apply to a bootcamp.  (You'll either need cybersource set up  if you want to try out payments)
- Check that `HubspotObjects` exist for your new `Contact`, `Order` and associated `Line`
- Log on to the hubspot website, go to the "bootcampsdev" sandbox account, click "Sales->Deals" from the top menu bar, and then search for your new orders (name should be "Bootcamp-application-order-<order_id>").  There may be multiple matches due to syncing from different local instances, but one of them should have the correct contact, price, product, etc.


#### Where should the reviewer start?
https://github.com/mitodl/hq/discussions/292

#### Any background context you want to provide?
This is a little hard to test thoroughly.  The CRM API doesn't allow for new products to be created on a standard non-enterprise hubspot account, which is what has been used for RC.  So I had to create a new sandbox account under the main bootcamps account for testing going forward, once this PR is released to RC.   I then did a dump of the RC database, restored to my local db, and used the management commands in this PR to sync to the new sandbox account.

The Hubspot API is [rate limited](https://legacydocs.hubspot.com/apps/api_guidelines) (150 calls/10 seconds), so for batch functions I used a combo of chunking and delays to try avoiding that max.
